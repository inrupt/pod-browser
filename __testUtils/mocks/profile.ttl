@prefix : <#>.
@prefix solid: <http://www.w3.org/ns/solid/terms#>.
@prefix pro: <./>.
@prefix n0: <http://xmlns.com/foaf/0.1/>.
@prefix schem: <http://schema.org/>.
@prefix n1: <http://www.w3.org/ns/auth/acl#>.
@prefix n2: <https://arquisoft.github.io/>.
@prefix ldp: <http://www.w3.org/ns/ldp#>.
@prefix inbox: </inbox/>.
@prefix sp: <http://www.w3.org/ns/pim/space#>.
@prefix meg: </>.
@prefix c: <https://megoth.solid.community/profile/card#>.
@prefix n: <http://www.w3.org/2006/vcard/ns#>.

pro:card a n0:PersonalProfileDocument; n0:maker :me; n0:primaryTopic :me .

:me
    a schem:Person, n0:Person;
    n0:name "Test Testersen";
    sp:preferencesFile </settings/prefs.ttl>;
    sp:storage meg:;
    solid:account meg:;
    solid:privateTypeIndex </settings/privateTypeIndex.ttl>;
    solid:publicTypeIndex </settings/publicTypeIndex.ttl> .
