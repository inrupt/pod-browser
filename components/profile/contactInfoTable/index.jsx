/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import T from "prop-types";
import React, { useContext, useState } from "react";
import { rdf, vcard } from "rdf-namespaces";
import clsx from "clsx";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import {
  addUrl,
  asUrl,
  createThing,
  getSourceUrl,
  getThing,
  getUrlAll,
  removeUrl,
  saveSolidDatasetAt,
  setThing,
  setUrl,
} from "@inrupt/solid-client";

import {
  DatasetContext,
  Table,
  TableColumn,
  useSession,
  useThing,
  Value,
} from "@inrupt/solid-ui-react";

import {
  Box,
  MenuItem,
  Typography,
  FormControl,
  Select,
  createStyles,
} from "@material-ui/core";

import { Table as PrismTable } from "@inrupt/prism-react-components";
import styles from "./styles";
import { chain } from "../../../src/solidClientHelpers/utils";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

const TESTCAFE_ID_DELETE_BUTTON = "profile-address-delete-button";
const TESTCAFE_ID_EXISTING_VALUE = "profile-address-existing-value";
const TESTCAFE_ID_NEW_BUTTON = "profile-address-new-button";
const TESTCAFE_ID_NEW_TYPE = "profile-address-new-type";
const TESTCAFE_ID_NEW_VALUE = "profile-address-new-value";

const CONTACT_TYPE_LABEL_MAP = {
  [vcard.Home]: "Home",
  [vcard.Work]: "Work",
};

export const PREFIX_MAP = {
  [vcard.hasTelephone]: "tel:",
  [vcard.hasEmail]: "mailto:",
};

export const DEFAULT_CONTACT_TYPE = vcard.Home;

export function setupAddContactDetail(
  property,
  newContactType,
  newContactValue,
  dataset,
  profile,
  saveHandler,
  setNewContactValue
) {
  return async () => {
    const prefix = PREFIX_MAP[property];
    const newContactDetail = chain(
      setUrl(createThing(), rdf.type, newContactType),
      (t) => setUrl(t, vcard.value, `${prefix}${newContactValue}`)
    );

    const datasetWithContactDetail = setThing(dataset, newContactDetail);

    const newProfile = addUrl(
      profile,
      property,
      asUrl(newContactDetail, getSourceUrl(dataset))
    );

    await saveHandler(newProfile, datasetWithContactDetail);

    setNewContactValue("");
  };
}

export function setupSaveHandler(fetch, setDataset) {
  return async (newThing, datasetToUpdate) => {
    const savedDataset = await saveSolidDatasetAt(
      getSourceUrl(datasetToUpdate),
      setThing(datasetToUpdate, newThing),
      { fetch }
    );
    setDataset(savedDataset);
  };
}

export function setupRemoveRow(profile, property, saveHandler, dataset) {
  return async (rowThing) => {
    const contactDetailUrl = asUrl(rowThing);
    const newProfile = removeUrl(profile, property, contactDetailUrl);
    await saveHandler(newProfile, dataset);
  };
}

export function setupDeleteButtonCell(editing, removeRow, bem) {
  return () => {
    const { thing: rowThing } = useThing();

    if (!editing) {
      return null;
    }

    return (
      <button
        onClick={() => removeRow(rowThing)}
        className={bem("button", "action")}
        type="button"
        data-testid={TESTCAFE_ID_DELETE_BUTTON}
      >
        Delete
      </button>
    );
  };
}

export default function ContactTable({ editing, property }) {
  const tableClass = PrismTable.useTableClass("table", "inherits");
  const classes = useStyles();
  const bem = useBem(classes);

  const [newContactType, setNewContactType] = useState(DEFAULT_CONTACT_TYPE);
  const [newContactValue, setNewContactValue] = useState("");
  const { fetch } = useSession();
  const { dataset, setDataset } = useContext(DatasetContext);
  const { thing: profile } = useThing();

  const contactDetailUrls = getUrlAll(profile, property);
  const contactDetailThings = contactDetailUrls.map((url) => ({
    dataset,
    thing: getThing(dataset, url),
  }));

  const saveHandler = setupSaveHandler(fetch, setDataset);

  const addContactDetail = setupAddContactDetail(
    property,
    newContactType,
    newContactValue,
    dataset,
    profile,
    saveHandler,
    setNewContactValue
  );

  const removeRow = setupRemoveRow(profile, property, saveHandler, dataset);

  const DeleteButtonCell = setupDeleteButtonCell(editing, removeRow, bem);

  return (
    <>
      <Box mb={1}>
        <Table
          things={contactDetailThings}
          className={clsx(tableClass, bem("table"))}
          getRowProps={() => ({
            className: clsx(bem("table__body-row")),
          })}
        >
          <TableColumn
            property={rdf.type}
            body={({ value }) => {
              const existingTypeValue = CONTACT_TYPE_LABEL_MAP[value] || value;
              return <Typography title={value}>{existingTypeValue}</Typography>;
            }}
            dataType="url"
            header={() => null}
          />

          <TableColumn
            property={vcard.value}
            body={() => (
              <Typography>
                <Value
                  edit={editing}
                  autosave
                  dataType="url"
                  inputProps={{
                    className: bem("input"),
                    "data-testid": TESTCAFE_ID_EXISTING_VALUE,
                  }}
                  property={vcard.value}
                  onSave={(savedDataset) => {
                    setDataset(savedDataset);
                  }}
                />
              </Typography>
            )}
            dataType="url"
            header={() => null}
          />

          <TableColumn
            property={vcard.value}
            body={DeleteButtonCell}
            dataType="url"
            header=""
          />
        </Table>
      </Box>

      {editing && (
        <>
          <Box display="flex" alignItems="center">
            <Box>
              <FormControl className={classes.formControl} variant="outlined">
                <Select
                  value={newContactType}
                  onChange={(e) => setNewContactType(e.target.value)}
                  inputProps={{
                    "data-testid": TESTCAFE_ID_NEW_TYPE,
                  }}
                >
                  {Object.keys(CONTACT_TYPE_LABEL_MAP).map((iri) => (
                    <MenuItem key={iri} value={iri}>
                      {CONTACT_TYPE_LABEL_MAP[iri]}
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </Box>

            <Box width="100%" ml={1} mr={1}>
              <input
                label="Value"
                value={newContactValue}
                onChange={(e) => setNewContactValue(e.target.value)}
                className={bem("input")}
                data-testid={TESTCAFE_ID_NEW_VALUE}
              />
            </Box>

            <Box width="8.25em" textAlign="center">
              <button
                type="button"
                onClick={addContactDetail}
                className={bem("button", "action")}
                data-testid={TESTCAFE_ID_NEW_BUTTON}
              >
                Add
              </button>
            </Box>
          </Box>
        </>
      )}
    </>
  );
}

ContactTable.propTypes = {
  editing: T.bool,
  property: T.string.isRequired,
};

ContactTable.defaultProps = {
  editing: false,
};
