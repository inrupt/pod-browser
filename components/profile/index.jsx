/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useState } from "react";
import T from "prop-types";
import { vcard } from "rdf-namespaces";
import Link from "next/link";
import {
  Avatar,
  Box,
  Paper,
  InputLabel,
  createStyles,
} from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import { Container, BackToNavLink } from "@inrupt/prism-react-components";

import { CombinedDataProvider, Text, Image } from "@inrupt/solid-ui-react";

import ContactInfoTable from "./contactInfoTable";
import styles from "./styles";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export default function Profile(props) {
  const { profileIri, editing } = props;
  const [error, setError] = useState(null);

  const classes = useStyles();
  const bem = useBem(classes);

  // TODO replace with toast error or something?
  if (error) {
    return error.toString();
  }

  return (
    <Container>
      <Box mt={2} mb={2}>
        <Link href="/contacts" passHref>
          <BackToNavLink>contacts</BackToNavLink>
        </Link>
      </Box>

      <Paper style={{ maxWidth: "800px" }}>
        <Box p={2}>
          <CombinedDataProvider
            datasetUrl={profileIri}
            thingUrl={profileIri}
            onError={setError}
          >
            <Box alignItems="center" display="flex">
              <Box>
                <Avatar className={classes.avatar}>
                  <Image
                    property={vcard.hasPhoto}
                    width={120}
                    errorComponent={() => (
                      <Avatar
                        className={bem("avatar")}
                        alt="Contact photo placeholder"
                      />
                    )}
                  />
                </Avatar>
              </Box>

              <Box p={2}>
                <h3>
                  <Text property={vcard.fn} />
                </h3>
              </Box>
            </Box>

            <hr />

            <Box mt={2}>
              <Box>
                <InputLabel>Name</InputLabel>
                <Text
                  property={vcard.fn}
                  edit={editing}
                  autosave
                  inputProps={{ className: bem("input") }}
                />
              </Box>

              <Box mt={1}>
                <InputLabel>Role</InputLabel>
                <Text
                  property={vcard.role}
                  edit={editing}
                  inputProps={{ className: bem("input") }}
                  autosave
                />
              </Box>

              <Box mt={1}>
                <InputLabel>Company</InputLabel>
                <Text
                  property={vcard.org}
                  edit={editing}
                  inputProps={{ className: bem("input") }}
                  autosave
                />
              </Box>
            </Box>

            <Box mt={4}>
              <InputLabel>Email Addresses</InputLabel>
              <ContactInfoTable property={vcard.hasEmail} editing={editing} />
            </Box>

            <Box mt={4}>
              <InputLabel>Phone Numbers</InputLabel>
              <ContactInfoTable
                property={vcard.hasTelephone}
                editing={editing}
              />
            </Box>
          </CombinedDataProvider>
        </Box>
      </Paper>
    </Container>
  );
}

Profile.propTypes = {
  profileIri: T.string.isRequired,
  editing: T.bool,
};

Profile.defaultProps = {
  editing: false,
};
