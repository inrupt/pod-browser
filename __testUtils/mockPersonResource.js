/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { addStringNoLocale, addUrl, mockThingFrom } from "@inrupt/solid-client";
import { vcard, foaf } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { packageProfile } from "../src/solidClientHelpers/profile";

export const aliceWebIdUrl = "http://example.com/alice#me";
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://example.com/alice.jpg";

export function mockPersonDatasetAlice() {
  return chain(
    mockThingFrom(aliceWebIdUrl),
    (t) => addStringNoLocale(t, vcard.fn, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto)
  );
}

export function mockProfileAlice() {
  return packageProfile(aliceWebIdUrl, mockPersonDatasetAlice());
}

export const bobWebIdUrl = "http://example.com/bob#me";
export const bobName = "Bob";
export const bobNick = "B";

export function mockPersonDatasetBob() {
  return chain(
    mockThingFrom(bobWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, bobName),
    (t) => addStringNoLocale(t, foaf.nick, bobNick)
  );
}

export function mockProfileBob() {
  return packageProfile(bobWebIdUrl, mockPersonDatasetBob());
}
