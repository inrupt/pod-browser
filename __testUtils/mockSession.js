/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import mockFetch from "./mockFetch";

import profile from "./mocks/profile.ttl";
import storageAclTtl from "./mocks/storageAcl.ttl";
import mockResponse from "./mockResponse";

export const webId = "http://example.com/webId#me";
export const storage = "http://example.com/";
export const storageAcl = "http://example.com/.acl";

export { profile };

export default function mockSession(options = {}) {
  // this instance is a normal, authenticated session
  // you can override the properties with what you need, but if you repeat something a lot
  // consider naming it and create a reusable mock function
  return {
    fetch: mockFetch({
      [webId]: () => mockResponse(200, profile),
      [storage]: () =>
        mockResponse(200, "", {
          Link: `<${storageAcl}>; rel="acl"`,
          url: storage,
        }),
      [storageAcl]: () =>
        mockResponse(200, storageAclTtl, {
          url: storageAcl,
        }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId,
    }, // add more properties as needed
    ...options,
  }; // add more properties as needed
}

// create other functions in this file to reflect other types of sessions we need to test with

export function mockUnauthenticatedSession() {
  return {
    fetch: () => mockFetch(),
    info: {
      isLoggedIn: false,
      sessionId: null,
      webId: null,
    }, // add more properties as needed
  }; // add more properties as needed
}

export function mockAuthenticatedSessionWithNoAccessToPod() {
  return {
    fetch: mockFetch({
      [webId]: () => mockResponse(200, profile),
      [storage]: () =>
        mockResponse(200, "", {
          Link: `<${storageAcl}>; rel="acl"`,
          url: storage,
        }),
      [storageAcl]: () =>
        mockResponse(401, "", {
          url: storageAcl,
        }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId,
    },
  };
}
