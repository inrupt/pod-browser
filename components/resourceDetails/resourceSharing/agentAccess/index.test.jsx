/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { mockSolidDatasetFrom } from "@inrupt/solid-client";
import { DatasetProvider } from "@inrupt/solid-ui-react";
import { waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AgentAccess, { getDialogId, saveHandler, submitHandler } from "./index";

import mockSessionContextProvider from "../../../../__testUtils/mockSessionContextProvider";
import mockSession from "../../../../__testUtils/mockSession";

import { renderWithTheme } from "../../../../__testUtils/withTheme";
import { createAccessMap } from "../../../../src/solidClientHelpers/permissions";
import useFetchProfile from "../../../../src/hooks/useFetchProfile";
import * as profileFns from "../../../../src/solidClientHelpers/profile";
import { mockProfileAlice } from "../../../../__testUtils/mockPersonResource";

jest.mock("../../../../src/solidClientHelpers/permissions");
jest.mock("../../../../src/hooks/useFetchProfile");

const webId = "http://example.com/webId#me";

describe("AgentAccess", () => {
  const permission = {
    acl: createAccessMap(),
    webId,
  };
  const datasetUrl = "http://example.com/dataset";
  const dataset = mockSolidDatasetFrom(datasetUrl);

  beforeEach(() => {
    useFetchProfile.mockReturnValue({ data: mockProfileAlice() });
  });

  it("renders", () => {
    const { asFragment } = renderWithTheme(
      <DatasetProvider dataset={dataset}>
        <AgentAccess permission={permission} />
      </DatasetProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("fetches profile for webId", () => {
    renderWithTheme(
      <DatasetProvider dataset={dataset}>
        <AgentAccess permission={permission} />
      </DatasetProvider>
    );
    expect(useFetchProfile).toHaveBeenCalledWith(webId);
  });

  it("renders an error message with a 'try again' button if it's unable to load profile", () => {
    useFetchProfile.mockReturnValue({ error: "error" });
    const { asFragment, getByTestId } = renderWithTheme(
      <DatasetProvider dataset={dataset}>
        <AgentAccess permission={permission} />
      </DatasetProvider>
    );
    expect(getByTestId("try-again-button")).toBeTruthy();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a spinner after clicking 'try again' button", async () => {
    useFetchProfile.mockReturnValue({ error: "error" });
    const { getByTestId } = renderWithTheme(
      <DatasetProvider dataset={dataset}>
        <AgentAccess permission={permission} />
      </DatasetProvider>
    );
    const button = getByTestId("try-again-button");
    userEvent.click(button);

    expect(getByTestId("try-again-spinner")).toBeTruthy();
  });

  it("tries to fetch the profile again when clicking 'try again' button", async () => {
    useFetchProfile.mockReturnValue({ error: "error" });
    const fetchProfile = jest.spyOn(profileFns, "fetchProfile");
    const { getByTestId } = renderWithTheme(
      <DatasetProvider dataset={dataset}>
        <AgentAccess permission={permission} />
      </DatasetProvider>
    );
    const button = getByTestId("try-again-button");
    userEvent.click(button);

    await waitFor(() =>
      expect(fetchProfile).toHaveBeenCalledWith(webId, expect.anything())
    );
  });

  it("removes the spinner when fetching succeeds", async () => {
    useFetchProfile.mockReturnValue({ error: "error" });
    const fetchProfile = jest.spyOn(profileFns, "fetchProfile");
    fetchProfile.mockReturnValue("profile");
    const { getByTestId, queryByTestId } = renderWithTheme(
      <DatasetProvider dataset={dataset}>
        <AgentAccess permission={permission} />
      </DatasetProvider>
    );
    const button = getByTestId("try-again-button");
    userEvent.click(button);

    await waitFor(() => expect(queryByTestId("try-again-spinner")).toBeFalsy());
  });

  it("removes the spinner when fetching errors", async () => {
    useFetchProfile.mockReturnValue({ error: "error" });
    const fetchProfile = jest.spyOn(profileFns, "fetchProfile");
    fetchProfile.mockReturnValue("profile");
    const { getByTestId, queryByTestId } = renderWithTheme(
      <DatasetProvider dataset={dataset}>
        <AgentAccess permission={permission} />
      </DatasetProvider>
    );
    const button = getByTestId("try-again-button");
    userEvent.click(button);

    await waitFor(() =>
      expect(fetchProfile).toHaveBeenCalledWith(webId, expect.anything())
    );
    await waitFor(() => expect(queryByTestId("try-again-spinner")).toBeFalsy());
  });

  it("renders a 'remove' button if it's unable to load profile", () => {
    useFetchProfile.mockReturnValue({ error: "error" });
    const { asFragment, getByTestId } = renderWithTheme(
      <DatasetProvider dataset={dataset}>
        <AgentAccess permission={permission} />
      </DatasetProvider>
    );
    expect(getByTestId("remove-button")).toBeTruthy();
    expect(asFragment()).toMatchSnapshot();
  });

  describe("user tries to change access for themselves", () => {
    it("checkboxes are disabled", () => {
      const session = mockSession();
      const SessionProvider = mockSessionContextProvider(session);

      const { asFragment } = renderWithTheme(
        <SessionProvider>
          <DatasetProvider dataset={dataset}>
            <AgentAccess permission={permission} webId={session.info.webId} />
          </DatasetProvider>
        </SessionProvider>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

describe("getDialogId", () => {
  it("generates dialogId", () =>
    expect(getDialogId("foo")).toEqual("change-agent-access-foo"));
});

describe("submitHandler", () => {
  let event;

  beforeEach(() => {
    event = { preventDefault: jest.fn() };
  });

  test("user changes their own permissions", () => {
    const setOpen = jest.fn();
    const dialogId = "dialogId";
    submitHandler(42, 42, setOpen, dialogId)(event);

    expect(event.preventDefault).toHaveBeenCalledWith();
    expect(setOpen).toHaveBeenCalledWith(dialogId);
  });
  test("user changes someone else's permissions", () => {
    const savePermissions = jest.fn();
    const tempAccess = "tempAccess";
    submitHandler(42, 1337, null, null, savePermissions, tempAccess)(event);

    expect(event.preventDefault).toHaveBeenCalledWith();
    expect(savePermissions).toHaveBeenCalledWith(tempAccess);
  });
});

describe("saveHandler", () => {
  const accessControl = {
    savePermissionsForAgent: jest.fn().mockResolvedValue({}),
  };
  let onLoading;
  let setAccess;
  let setTempAccess;
  let setSeverity;
  let setMessage;
  let setAlertOpen;
  const newAccess = "newAccess";
  let savePermissions;

  beforeEach(() => {
    onLoading = jest.fn();
    setAccess = jest.fn();
    setTempAccess = jest.fn();
    setSeverity = jest.fn();
    setMessage = jest.fn();
    setAlertOpen = jest.fn();
    savePermissions = saveHandler(
      accessControl,
      onLoading,
      setAccess,
      webId,
      setTempAccess,
      setSeverity,
      setMessage,
      setAlertOpen
    );
  });

  test("save is successful", async () => {
    await expect(savePermissions(newAccess)).resolves.toBeUndefined();

    expect(onLoading).toHaveBeenCalledWith(true);
    expect(setAccess).toHaveBeenCalledWith(newAccess);
    expect(accessControl.savePermissionsForAgent).toHaveBeenCalledWith(
      webId,
      newAccess
    );
    expect(setTempAccess).toHaveBeenCalledWith(null);
    expect(setSeverity).toHaveBeenCalledWith("success");
    expect(setMessage).toHaveBeenCalledWith("Permissions have been updated!");
    expect(setAlertOpen).toHaveBeenCalledWith(true);
    expect(onLoading).toHaveBeenCalledWith(false);
  });

  test("save request fails", async () => {
    const error = "error";
    accessControl.savePermissionsForAgent.mockResolvedValue({ error });

    await expect(savePermissions(newAccess)).rejects.toEqual(error);

    expect(onLoading).toHaveBeenCalledWith(true);
    expect(setAccess).toHaveBeenCalledWith(newAccess);
    expect(accessControl.savePermissionsForAgent).toHaveBeenCalledWith(
      webId,
      newAccess
    );
  });
});
