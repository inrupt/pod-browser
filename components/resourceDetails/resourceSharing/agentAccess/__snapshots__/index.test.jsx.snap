// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentAccess for authenticated agent renders correctly 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-nameAndAvatarContainer"
  >
    <div
      class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
    >
      <span
        class="PodBrowser-icon-user-lock PodBrowser-avatar__icon PodBrowser-avatar__icon--authenticated"
      />
    </div>
    <p
      class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
      data-testid="agent-web-id"
    />
    <button
      class="PodBrowser-button"
      data-testid="menu-button"
      type="button"
    >
      <i
        class="PodBrowser-icon-more"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess for public agent renders correctly 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-nameAndAvatarContainer"
  >
    <div
      class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
    >
      <span
        class="PodBrowser-icon-globe PodBrowser-avatar__icon PodBrowser-avatar__icon--public"
      />
    </div>
    <p
      class="PodBrowser-root PodBrowser-detailText PodBrowser-body1 PodBrowser-detailText"
      data-testid="agent-web-id"
    />
    <button
      class="PodBrowser-button"
      data-testid="menu-button"
      type="button"
    >
      <i
        class="PodBrowser-icon-more"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess with profile renders 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loadingStateContainer"
  >
    <span
      class="PodBrowser-root PodBrowser-circle PodBrowser-pulse"
      style="width: 40px; height: 40px;"
    />
    <div>
      <span
        class="PodBrowser-root PodBrowser-text PodBrowser-pulse"
        style="width: 100px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess without profile renders an error message with a 'try again' button if it's unable to load profile 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loadingStateContainer"
  >
    <span
      class="PodBrowser-root PodBrowser-circle PodBrowser-pulse"
      style="width: 40px; height: 40px;"
    />
    <div>
      <span
        class="PodBrowser-root PodBrowser-text PodBrowser-pulse"
        style="width: 100px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess without profile renders skeleton placeholders when profile is not available 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loadingStateContainer"
  >
    <span
      class="PodBrowser-root PodBrowser-circle PodBrowser-pulse"
      style="width: 40px; height: 40px;"
    />
    <div>
      <span
        class="PodBrowser-root PodBrowser-text PodBrowser-pulse"
        style="width: 100px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`AgentAccess without profile returns null when no access 1`] = `<DocumentFragment />`;
